docs/custom_promise_types/gpg_keys_using_lib.py

```diff
@@ -0,0 +1,45 @@
+import sys
+import os
+from cfengine import PromiseModule, ValidationError, Result
+
+
+class GpgKeysPromiseTypeModule(PromiseModule):
+    def __init__(self):
+        super().__init__("gpg_keys_promise_module", "0.0.1")
+
+    def validate_promise(self, promiser, attributes):
+      if not promiser.startswith("/"):
+        raise ValidationError(f"GNUPGHOME path '{promiser}' must be absolute")
+
+    def evaluate_promise(self, promiser, attributes):
+      keydir = attributes["keydir"]
+      result = Result.KEPT
+
+# TODO handle errors with os calls and such
+      with os.scandir(keydir) as it:
+        for entry in it:
+          if entry.name.endswith('.fp') and entry.is_file():
+            fp_filename = keydir + os.sep + entry.name
+            fp_file = open(fp_filename, "r")
+            fp = fp_file.read()
+            fp = fp.replace('\n','')
+            fp_file.close()
+            self.log_debug(f"file {fp_filename} has fingerprint {fp}")
+            rc = os.system(f"gpg --homedir {promiser} -k {fp} 2>/dev/null >/dev/null")
+            if rc != 0:
+              asc_file = keydir + os.sep + entry.name.replace(".fp",".asc")
+              self.log_info(f"Importing {asc_file} public key file with fingerprint {fp} into GNUPGHOME {promiser}")
+              rc = os.system(f"gpg --homedir {promiser} --import {asc_file} 2>/dev/null >/dev/null")
+              if rc != 0:
+                self.log_error(f"Unable to import key file {asc_file} for fingerprint {fp}, exit code {rc}")
```

Comment:
I like to put single quotes around strings of data in log messages. It is especially useful if that variable might contain a newline, space, tab or something similar. Makes it much easier to see where it starts and ends.